backend/tests/v5/e2e/routes/teamspaces/roles.test.js

```diff
@@ -0,0 +1,214 @@
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+const { times } = require('lodash');
+const SuperTest = require('supertest');
+const ServiceHelper = require('../../../helper/services');
+const { src } = require('../../../helper/path');
+const { generateRandomString } = require('../../../helper/services');
+
+const { DEFAULT_ROLES } = require(`${src}/models/roles.constants`);
+const { templates } = require(`${src}/utils/responseCodes`);
+
+let server;
+let agent;
+
+const generateBasicData = () => {
+	const users = {
+		tsAdmin: ServiceHelper.generateUserCredentials(),
+		normalUser: ServiceHelper.generateUserCredentials(),
+		nobody: ServiceHelper.generateUserCredentials(),
+	};
+
+	const teamspace = ServiceHelper.generateRandomString();
+
+	const roles = times(10, () => ({
+		_id: ServiceHelper.generateUUIDString(),
+		name: ServiceHelper.generateRandomString(),
+		color: ServiceHelper.generateRandomString(),
+	}));
+
+	return {
+		users,
+		teamspace,
+		roles,
+	};
+};
+
+const setupData = async ({ users, teamspace, roles }) => {
+	await ServiceHelper.db.createTeamspace(teamspace, [users.tsAdmin.user]);
+
+	await Promise.all([
+		ServiceHelper.db.createUser(users.tsAdmin, [teamspace]),
+		ServiceHelper.db.createUser(users.normalUser, [teamspace]),
+		ServiceHelper.db.createUser(users.nobody, []),
+		ServiceHelper.db.createRoles(teamspace, roles),
+	]);
+};
+
+const testGetRoleList = () => {
+	const basicData = generateBasicData();
+	const { users, teamspace, roles } = basicData;
+	beforeAll(async () => {
+		await setupData(basicData);
+	});
+
+	const route = (key, ts = teamspace) => `/v5/teamspaces/${ts}/roles${key ? `?key=${key}` : ''}`;
+
+	describe('Get role list', () => {
+		describe.each([
+			['user does not have a valid session', undefined, undefined, false, templates.notLoggedIn],
+			['teamspace does not exist', users.tsAdmin.apiKey, generateRandomString(), false, templates.teamspaceNotFound],
+			['user is not a member of the teamspace', users.nobody.apiKey, undefined, false, templates.teamspaceNotFound],
+			['user is a ts admin', users.tsAdmin.apiKey, undefined, true],
+			['user is a member', users.normalUser.apiKey, undefined, true],
+		])('', (desc, key, ts, success, expectedRes) => {
+			test(`should ${success ? 'succeed' : `fail with ${expectedRes.code}`} if ${desc}`, async () => {
+				const expectedStatus = success ? templates.ok.status : expectedRes.status;
+				const res = await agent.get(route(key, ts)).expect(expectedStatus);
+				if (success) {
+					expect(res.body.roles).toEqual(expect.arrayContaining(roles));
```

Comment:
There's an [out of order equal array check function](https://github.com/3drepo/3drepo.io/blob/master/backend/tests/v5/helper/services.js#L350) in service helper. would be more appropriate to use that as arrayContaining would only check for the subset